### ISSUE TITLE
Upgrade from 1.5.3 -> 2.0.2 breaks Flask-JWT-Extended

### ISSUE THREAD

**Issue by wittlesouth** [zalando/connexion#806]
### Description

I've been using Connexion 1.X with Flask-JWT-Extended for multiple years without issue. I recently started a new project, and decided to start with Connexion 2.0.2. Unfortunately, it appears that Connexion 2.0.2 breaks Flask-JWT-Extended. Every unit test I have that tests against a protected endpoint fails with 2.0.2, but works with 1.5.3 and below.

### Expected behaviour

Flask-JWT-Extended decorators should work for Connexion endpoints.

### Actual behaviour

Any HTTP requests sent to Flask-JWT-Extended protected endpoints fail with "No authentication token provided."

### Steps to reproduce

I have a Flask app that eposes a small set of REST APIs, and the project I'm building them in has a set of unit tests. The test process is to start an instance of the API server in a Kubernetes container, then run a set of tests that make requests to the APIs. In my local environment, the tests run from an OS X shell, and make http requests to the server running in a minikube container. In my Jenkins environment, the tests run from one container, and make http requests to the server in another container.

The issue is that the tests all pass in my local environment, but any test that runs against a flask-jwt-extended endpoint in my Jenkins environment fail due to missing access token errors from the flask-jwt-extended decorators. 

Here is the request handler method and decorator for the first test that fails with this symptom.

```
# The /login enpoint with the GET verb is intended to be used by client apps
# to reload the user's application data when needed (e.g. after an application
# refresh, or browser restart)
@jwt_refresh_token_required
def search():
    """Handles GET verb for /login endpoint"""
    return jsonify(hydrate(g.user)), 200
```

I have before_request and after_request logging enabled with the following code:

```
@FAPP.before_request
def before_request():
    """Method to do work before the request"""
    # Ensure there is a database session available for the request
    g.db_session = DBSESSION()

    # Log request headers & body if FLASK_DEBUG > 1
    if 'FLASK_DEBUG' in os.environ and int(os.environ['FLASK_DEBUG']) >= 2:
        LOGGER.debug('Starting request for path %s', request.url)
        LOGGER.debug('Request Headers: %s', request.headers)
        LOGGER.debug('Request Body: %s', request.get_data())
        LOGGER.debug('Request Cookies: ' + str(request.cookies))

```

and

```
@FAPP.after_request
def after_request(resp):
    """Method to do work after the request"""
    # Log request headers & body if FLASK_DEBUG > 1
    if 'FLASK_DEBUG' in os.environ and int(os.environ['FLASK_DEBUG']) >= 2:
        LOGGER.debug('Response Headers: %s', resp.headers)
        LOGGER.debug('Response Body: %s', resp.get_data())

```

Here is a snippet of the logs from the server in the Jenkins test that is failing

```
[2018-12-01 11:59:52,854] DEBUG in server: Starting request for path http://rftw-jenkins-test-server.wittlesouth:8280/api/v1/login
[2018-12-01 11:59:52,854] DEBUG in server: Request Headers: Accept: */*
X-Csrf-Token: 510f5491-8e2c-4c49-add9-b7bcf21de353
Cookie: refresh_token_cookie=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJuYmYiOjE1NDM2NjU1OTIsImV4cCI6MTU0NjI1NzU5MiwiaWRlbnRpdHkiOiI4ZjkyNGQ4OC0zNzUyLTQzYzYtODhiOC1mNjA0Yzc4MzAyZGMiLCJ0eXBlIjoicmVmcmVzaCIsImp0aSI6IjJjZjYxMTY1LTkxYzQtNDJmYS1iYTk1LWU0Njc0YjY1N2Y1MiIsImNzcmYiOiI1MTBmNTQ5MS04ZTJjLTRjNDktYWRkOS1iN2JjZjIxZGUzNTMiLCJpYXQiOjE1NDM2NjU1OTJ9.Zmaw-Zar1psqzVbGyKnttbQZW9zwbIuOytEkrKpNuYU; access_token_cookie=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJuYmYiOjE1NDM2NjU1OTIsImV4cCI6MTU0MzY2NzM5MiwidHlwZSI6ImFjY2VzcyIsImp0aSI6ImViMWFmNTA5LTRjOWUtNGQ1MS04Yzg1LWRlMzJmODNlOGU0NCIsImNzcmYiOiJmZjYyOTMyNC02OWI5LTQ0MGMtYTM5ZC1kMWQwNDdiMzFkNzMiLCJmcmVzaCI6ZmFsc2UsImlkZW50aXR5IjoiOGY5MjRkODgtMzc1Mi00M2M2LTg4YjgtZjYwNGM3ODMwMmRjIiwiaWF0IjoxNTQzNjY1NTkyfQ.p1_RVP0rhnzcRvhVxqjtMMa7wpCeM7Ymcvii5A3HN2w; csrf_refresh_token=510f5491-8e2c-4c49-add9-b7bcf21de353; csrf_access_token=ff629324-69b9-440c-a39d-d1d047b31d73
Host: rftw-jenkins-test-server.wittlesouth:8280
User-Agent: python-requests/2.20.1
Connection: keep-alive
Accept-Encoding: gzip, deflate


[2018-12-01 11:59:52,854] DEBUG in server: Request Body: b''
[2018-12-01 11:59:52,854] DEBUG in server: Request Cookies: {'csrf_refresh_token': '510f5491-8e2c-4c49-add9-b7bcf21de353', 'refresh_token_cookie': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJuYmYiOjE1NDM2NjU1OTIsImV4cCI6MTU0NjI1NzU5MiwiaWRlbnRpdHkiOiI4ZjkyNGQ4OC0zNzUyLTQzYzYtODhiOC1mNjA0Yzc4MzAyZGMiLCJ0eXBlIjoicmVmcmVzaCIsImp0aSI6IjJjZjYxMTY1LTkxYzQtNDJmYS1iYTk1LWU0Njc0YjY1N2Y1MiIsImNzcmYiOiI1MTBmNTQ5MS04ZTJjLTRjNDktYWRkOS1iN2JjZjIxZGUzNTMiLCJpYXQiOjE1NDM2NjU1OTJ9.Zmaw-Zar1psqzVbGyKnttbQZW9zwbIuOytEkrKpNuYU', 'access_token_cookie': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJuYmYiOjE1NDM2NjU1OTIsImV4cCI6MTU0MzY2NzM5MiwidHlwZSI6ImFjY2VzcyIsImp0aSI6ImViMWFmNTA5LTRjOWUtNGQ1MS04Yzg1LWRlMzJmODNlOGU0NCIsImNzcmYiOiJmZjYyOTMyNC02OWI5LTQ0MGMtYTM5ZC1kMWQwNDdiMzFkNzMiLCJmcmVzaCI6ZmFsc2UsImlkZW50aXR5IjoiOGY5MjRkODgtMzc1Mi00M2M2LTg4YjgtZjYwNGM3ODMwMmRjIiwiaWF0IjoxNTQzNjY1NTkyfQ.p1_RVP0rhnzcRvhVxqjtMMa7wpCeM7Ymcvii5A3HN2w', 'csrf_access_token': 'ff629324-69b9-440c-a39d-d1d047b31d73'}
[2018-12-01 11:59:52,855] DEBUG in server: Response Headers: Content-Type: application/problem+json
Content-Length: 119


[2018-12-01 11:59:52,855] DEBUG in server: Response Body: b'{\n  "detail": "No authorization token provided",\n  "status": 401,\n  "title": "Unauthorized",\n  "type": "about:blank"\n}\n'
10.47.0.13 - - [01/Dec/2018 11:59:52] "GET /api/v1/login HTTP/1.1" 401 -
```

Although the refresh-token-cookie required by the @jwt_refresh_token_required decorator is present in the request, the operation still fails with the error "No authorization token provided".

Here is the section of the server code where I set the flask JWT configuration.

```
FAPP.config['JWT_TOKEN_LOCATION'] = ['cookies']
if 'RFTW' in os.environ and os.environ['RFTW'] == 'PROD':
    FAPP.config['JWT_COOKIE_SECURE'] = True
else:
    FAPP.config['JWT_COOKIE_SECURE'] = False
FAPP.config['JWT_ACCESS_COOKIE_PATH'] = '/'
FAPP.config['JWT_ACCESS_TOKEN_EXPIRES'] = datetime.timedelta(minutes=30)
FAPP.config['JWT_REFRESH_COOKIE_PATH'] = os.environ['API_PATH']
FAPP.config['JWT_REFRESH_TOKEN_EXPIRES'] = datetime.timedelta(days=30)
FAPP.config['JWT_SESSION_COOKIE'] = False
FAPP.config['JWT_COOKIE_CSRF_PROTECT'] = True
FAPP.config['JWT_CSRF_METHODS'] = ['POST', 'PUT', 'PATCH', 'DELETE', 'GET']
```

The container specs for the server container are not identical for the local environment vs. the Jenkins environment, but base image and python-related packages installed are the same. Here are the key parts of the local test server container spec for each.

Local test environment (where the server application works correctly):

```
FROM ubuntu:16.04
LABEL maintainer="eric@wittlesouth.com" project="ourlifestories.net"
RUN apt-get update -y
RUN apt-get install -y python3-pip python3-dev build-essential mysql-client libmysqlclient-dev curl sudo
ADD requirements.txt /
RUN pip3 install -r /requirements.txt
CMD ["bash", "-c", "${SERVER_COMMAND}"]
```

Jenkins test environment (where the above issue happens):

```
FROM ubuntu:16.04
LABEL maintainer="eric@wittlesouth.com" project="Roadmaps FTW"
RUN apt-get update -y
RUN apt-get install -y python3-pip python3-dev build-essential mysql-client libmysqlclient-dev curl sudo
# Below doesn't seem to work to get time correct on server
# RUN sudo echo "America/New_York" > /etc/timezone
# RUN sudo dpkg-reconfigure -f noninteractive tzdata
ADD requirements.txt /
RUN pip3 install -r /requirements.txt
RUN mkdir /app
ADD src /app/
CMD ["bash", "-c", "${SERVER_COMMAND}"]
```

If I change the Connexion version to 1.5.3, everything works fine. If I use 2.0.2, all endpoints with Flask-JWT-Extended decorators return errors. Endpoints without decorators work fine.

### Additional info:

Output of the commands:

- `python --version` = 3.5.3
- `pip show connexion | grep "^Version\:"` = 2.0.2
- Flask-JWT-Extended version is 3.13.1

Let me know if any additional information would be beneficial.

**Comment by dtkav** [comment]
Hey @wittlesouth - I'm not super familiar with the auth parts in Connexion 2.0, but @krise3k may be able to help.
I think your issue might be from applying the decorator to the handler function, which would apply auth much too late (after the entire connexion layer).
There are docs on how to set your auth function here : https://github.com/zalando/connexion/blob/master/docs/security.rst
Hope that helps!

**Comment by ctippur** [comment]
Facing the same issue.

**Comment by wittlesouth** [comment]
@dtkav - Maybe I wasn't clear what I'm trying to do. I don't want to use any part of the Connexion 2.0 authorization right now; I have already implemented authorization using the flask-JWT-extended package. My authorization works fine with Connexion 1.5.3, and fails with 2.0.2. Therefore, I'm pretty sure that there is a conflict between the two packages. It hasn't irritated me enough yet to figure out what it is on my own, but it's getting close.

**Comment by dtkav** [comment]
hey @wittlesouth - I understand what you're trying to do. Sorry that it's frustrating.
One thing to note is that applying a security decorator to the handler function is fundamentally insecure.

In addition, Connexion uses semantic versioning, so 2.0.x is a major revision bump, and includes some backwards-incompatible changes. You can always pin connexion to 1.5.3.

If you still want to upgrade to connexion 2.0, and are ok with applying the security decorator in an insecure way, then I'm happy to help debug.

In the readme, under 'new in connexion 2.0', there's a section that reads:
> If unsupported security requirements are defined or x-tokenInfoFunc/x-tokenInfoUrl is missing, connexion now denies requests instead of allowing access without security-check.

Based on this, I think you may need to set x-tokenInfoFunc and bypass connexion auth.

It would help me to tease apart the two issues you're having (environment, version) - can you confirm that connexion 2.0 doesn't work for you in either environment?

**Comment by wittlesouth** [comment]
One more update. I went and read through the 2.0 security doc quickly, and checked out the sample app even more quickly. I would guess based on that quick read-through, the "unsupported security requirements are defined" statement is likely to refer to the API specification. I frankly forgot that my API specification does include some attempt to define the security approach for the API. As far as I know, in connexion versions prior to 2.0, this specification had no effect, and at best was useful for generated documentation.

Do you know if I remove the security specification from the API spec entirely, will connexion 2.0 revert to 1.X behavior, namely avoiding any attempt at authentication / security?

I took a quick look a the JWT example, from that example it doesn't appear that the JWT implementation in Connexion is nearly as rich as what Flask-JWT-Extended provides. Specifically, I'm using both authentication and refresh tokens (with different time to live values), as well as including the tokens in form posts for (hopefully) CSRF protection. It wasn't clear from my admittedly quick read that the Connexion JWT implementation supports either of these capabilities.

**Comment by dtkav** [comment]
>  I frankly forgot that my API specification does include some attempt to define the security approach for the API. As far as I know, in connexion versions prior to 2.0, this specification had no effect, and at best was useful for generated documentation.

The purpose of connexion is to automatically do authentication, validation, serialization/de-serialization based on this specification. The idea is that the spec defines the interface (which connection handles), and then you define the business logic in the handler function.
In connexion 1.5.3, the only supported method of auth was oauth2 (IIRC). It would pass through any other auth methods (without doing any auth). This behavior was changed in connexion 2.0 to require the user to provide an auth handler if they had specified an auth method. You can set this auth handler to a pass-through if you like.

> Can you elaborate, or provide a reference, as to how "applying a security decorator is fundamentally insecure"? That is the fundamental methodology of Flask-JWT-Extended and numerous other Flask extensions for authentication I reviewed before I selected a package.

The way connexion is implemented is basically by applying several decorators to each handler function. The connexion security decorator is applied first, so that the minimal amount of connexion code runs before the security decorator is applied.

If you apply the decorator to the handler function, what's basically happening is (pseudocode):
```
@connexion_security
@connexion_array_handling
@connexion_validation
@connexion_deserialization
@connexion_etc...
@jwt_refresh_token_required
def search():
    """Handles GET verb for /login endpoint"""
    return jsonify(hydrate(g.user)), 200
```
And all of that code is running before your `jwt_refresh_token_requried` decorator, so your attack surface is really large, and your users will only get an auth error if they provide a perfectly formatted payload (otherwise they'll get a validation error).

What you're doing in connexion 1.5.3 is this:
```
@connexion_security  # pass through - doesn't support JWT auth!
@connexion_array_handling
@connexion_validation
@connexion_deserialization
@connexion_etc...
@jwt_refresh_token_required
def search():
    """Handles GET verb for /login endpoint"""
    return jsonify(hydrate(g.user)), 200
```

what's happening in 2.0.x is this:
```
@connexion_security  # AUTH ERROR! Connexion 2.0.x is more strict!
@connexion_array_handling
@connexion_validation
@connexion_deserialization
@connexion_etc...
@jwt_refresh_token_required
def search():
    """Handles GET verb for /login endpoint"""
    return jsonify(hydrate(g.user)), 200
```

> What does "If unsupported security requirements are defined" mean? Are you talking about security requirements in the API specification, or somewhere else? I certainly don't think I intended to configure any security validation by Connexion, so I assumed that this statement would not apply to me.

If you have security definitions in your spec, then they are "defined" for the purpose of connexion.

> I took a quick look a the JWT example, from that example it doesn't appear that the JWT implementation in Connexion is nearly as rich as what Flask-JWT-Extended provides
> ...
> It wasn't clear from my admittedly quick read that the Connexion JWT implementation supports either of these capabilities.

Connexion doesn't implement JWT - but it does currently pull the Bearer token from the header and then passes it to the function that you've registered with `x-BearerInfoFunc` in the spec.

AFAICT for your use case there is a missing features required before you could register an equivalent jwt function with `x-BearerInfoFunc`:
a) ability to use cookies with JWT auth

There is some work going on in #791 to support cookies, but I'm not sure it will work with JWT.

After that, you should be able to write a `x-BearerInfoFunc` that just calls `flask_jwt_extended.verify_jwt_refresh_token_in_request()`.

**Comment by dtkav** [comment]
@krise3k please jump in if you have some ideas for how to implement cookies for JWT and/or passing multiple tokens, or if you have any other ideas on how to help @wittlesouth

**Comment by dtkav** [comment]
@cziebuhr do you have any thoughts / ideas on supporting JWT refresh tokens and cookies?

**Comment by cziebuhr** [comment]
Do you have `security` defined in your swagger.spec? If you remove that, connexion won't do any authentication/authorization and you should be able to use your own decorators.

**Comment by wittlesouth** [comment]
The answers above provide an extensive set of options for further investigation on my part. I need to do some more reading and an experiment or two. Unfortunately with this busy time of year, it is likely to be several days before I come back to this.

**Comment by wittlesouth** [comment]
OK, I was wrong about how long it would take me to come back. Some more reading has lead me to a somewhat different conclusion. According to the OpenAPI 3.0 specification, the only valid authentication types in the spec are defined ("apiKey", "http", "oauth2", "openIdConnect"). Most of these (except openIdConnect) seem to be directly supported by one of the methods defined in the Connexion 2.0 security section. It doesn't appear that the authentication schemes provided by Flask-JWT-Extended comply with any of the authentication protocols defined in the OpenAPI 3.0 spec. And that would lead me to the conclusion that if I want a OpenAPI 3.0 compliant specification, I need to omit the security definition component of the API spec.

Am I reading the OpenAPI docs correctly?

Flask-JWT-Extended is pretty customizable. In my implementation, there are four distinct authentication components, the first two of which are HTTP only:

- access_token_cookie - This essentially represents a session; identifies a user and has a ~30 minute timeout
- refresh_token_cookie - This identifies a user and has a much longer timeout, but requires a reload before making data changes
- csrf_access_token - required to be passed back to the server in a custom HTTP header for session-oriented requests
- csrf_refresh_token - required to be passed back to the server in a custom HTTP header for identity-oriented requests (that are accessible after an individual session has expired)

Successful back-end authentication varies by which decorator you use, but can be based on either of the two HTTP only cookies as well as a request header with one of the two csrf tokens.

Correctly documenting the security requirements of any of my API endpoints would require documenting which of the two HTTP only cookies is required, as well as which of the two CSRF tokens, if any, is required in the header. Since Flask-JWT-Extended allows customization of the cookie names, and HTTP header name, those would need to be able to be documented.

If the schemes provided by Flask-JWT-Extended are formalized in a IETF document, I'd be surprised.

If I've read the OpenAPI 3.0 docs correctly, please let me know, and I'll close this issue. I will then need to decide whether to continue with the authentication schema I have (with the security requirements documented outside the formal API docs), or convert my application to an IETF standard authentication approach supported by the OpenAPI 3.0 specification.

**Comment by wittlesouth** [comment]
I removed the security specification from my APIs, and I'm now able to run Connexion 2.1 without an error. Thank you for the detailed responses, that pointed me in the right direction. I don't know if this problem is going to trip up others, but it would have helped me if there was a debug log message or something in the error message response that showed that the missing token detection error was generated from Connexion. The error text was similar enough to what Flask-JWT-Extended generates when its decorators detect a missing token that I looked there first.
